s3storagebasedauthorizationmanager/src/main/java/org/sfdc/uip/hive/ql/security/authorization/S3StorageBasedAuthorizationProvider.java

```diff
@@ -0,0 +1,316 @@
+            if (actions.isEmpty()) {
+                return;
+            }
+            checkPermissions(path, actions);
+        } catch (AccessControlException ex) {
+            throw authorizationException(ex);
+        } catch (Exception e) {
+            throw new HiveAccessControlException("Failed to authorize request. ", e);
+        }
+    }
+
+    /**
+     * Given a Privilege[], find out what all S3Actions are required
+     */
+    protected EnumSet<S3Action> getS3Actions(Privilege[] privs) {
+        EnumSet<S3Action> actions = EnumSet.noneOf(S3Action.class);
+
+        if (privs == null) {
+            return actions;
+        }
+
+        for (Privilege priv : privs) {
+            actions.add(getS3Action(priv));
+        }
+
+        return actions;
+    }
+
+    /**
+     * Given a privilege, return what S3Actions are required
+     */
+    protected S3Action getS3Action(Privilege priv) {
+
+        switch (priv.getPriv()) {
+            case ALL:
+                return S3Action.ALL;
+            case ALTER_DATA:
+            case ALTER_METADATA:
+            case CREATE:
+            case DROP:
+                return S3Action.WRITE;
+            case INDEX:
+                throw new AuthorizationException(
+                        "StorageBasedAuthorizationProvider cannot handle INDEX privilege");
+            case LOCK:
+                throw new AuthorizationException(
+                        "StorageBasedAuthorizationProvider cannot handle LOCK privilege");
+            case SELECT:
+            case SHOW_DATABASE:
+                return S3Action.READ;
+            case UNKNOWN:
+            default:
+                throw new AuthorizationException("Unknown privilege: " + priv.toString());
+        }
+    }
+
+    private void checkPermissions(String path, EnumSet<S3Action> actions) throws AccessControlException {
+        String userName = this.authenticator.getUserName();
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format("Checking permissions for user: %s for path: %s for actions: %s", userName, path, actions.toString()));
+        }
+
+        AWSCredentials sessionCredentials = urmCredentialsRetriever.getCredentialsForUser(userName);
+        AmazonS3 s3Client = AmazonS3ClientBuilder.standard()
+                    .withCredentials(new AWSStaticCredentialsProvider(sessionCredentials))
+                    .build();
+
+        for(S3Action action : actions){
+            checkActionS3(s3Client, action, path, userName);
+        }
+    }
+
+    private void checkActionS3(AmazonS3 s3Client, S3Action action, String path, String userName) throws AccessControlException  {
+        AmazonS3URI s3URIparser = new AmazonS3URI(path);
+
+        String bucketName = s3URIparser.getBucket();
+        String objectKey = s3URIparser.getKey();
+        String prefix = objectKey + "/";
+
+        if (action.equals(S3Action.READ)) {
+            //Allow read permission to all
```

Comment:
In the future, we may add a flag here so that authorization is actually done at the storage level for Reads as well. We think that this would be expected behavior from other users of URM.